src/course-unit/course-xblock/xblock-content/XBlockContent.jsx

```diff
@@ -1,68 +1,153 @@
+ *
+ * We use an IFrame so that the XBlock code, including user-authored HTML,
+ * cannot access things like the user's cookies, nor can it make GET/POST
+ * requests as the user. However, it is allowed to call any XBlock handlers.
+ */
 const XBlockContent = ({
-  id,
-  iframeUrl,
-  elementId,
-  onLoaded,
-  title,
+  view, type, getHandlerUrl, onBlockNotification,
 }) => {
-  const intl = useIntl();
-  const {
-    iframeHeight,
-    handleIFrameLoad,
-    showError,
-    hasLoaded,
-  } = useIFrameBehavior({
-    elementId, id, iframeUrl, onLoaded,
-  });
-
-  const iframeProps = {
-    id: elementId,
-    src: iframeUrl,
-    allow: IFRAME_FEATURE_POLICY,
-    frameBorder: 0,
-    allowFullScreen: true,
-    height: iframeHeight,
-    scrolling: 'no',
-    referrerPolicy: 'origin',
-    'data-testid': 'content-iframe-test-id',
-    onLoad: handleIFrameLoad,
-  };
+  const iframeRef = useRef(null);
+  const [html, setHtml] = useState(null);
+  const [iFrameHeight, setIFrameHeight] = useState(0);
+  const [iframeKey, setIframeKey] = useState(0);
+  const [isLoading, setIsLoading] = useState(true);
+
+  useEffect(() => {
+    const processView = () => {
+      if (view.html) {
+        const newHtml = wrapBlockHtmlForIFrame(
+          view.html,
+          view.resources,
+          getConfig().STUDIO_BASE_URL,
+          type,
+        );
+
+        // Load the XBlock HTML into the IFrame:
+        // iframe will only re-render in React when its property changes (key here)
+        setHtml(newHtml);
+        setIframeKey(prevKey => prevKey + 1);
+      }
+    };
+
+    // Process the XBlock view:
+    processView();
+  }, [view, type]);
+
+  useEffect(() => {
+    // Handle any messages we receive from the XBlock Runtime code in the IFrame.
+    // See iframe-wrapper.js to see the code that sends these messages.
+    const receivedWindowMessage = async (event) => {
+      if (iframeRef.current === null || event.source !== iframeRef.current.contentWindow) {
+        return;
+      } // This is some other random message.
+
+      const { method, replyKey, ...args } = event.data;
+
+      const frame = iframeRef.current.contentWindow;
+
+      const sendReply = async (data) => {
+        frame.postMessage({ ...data, replyKey }, '*');
+      };
+
+      if (method === 'bootstrap') {
+        await sendReply({ initialHtml: html });
+      } else if (method === 'get_handler_url') {
+        const handlerUrl = await getHandlerUrl(args.usageId);
+        await sendReply({ handlerUrl });
+      } else if (method === 'update_frame_height') {
+        setIFrameHeight(args.height);
+      } else if (method?.indexOf('xblock:') === 0) {
+        if (onBlockNotification) {
+          // This is a notification from the XBlock's frontend via 'runtime.notify(event, args)'
+          onBlockNotification({
+            eventType: method.substr(7), // Remove the 'xblock:' prefix that we added in iframe-wrapper.ts
```

Comment:
```suggestion
            eventType: method.substr('xblock'.length), // Remove the 'xblock:' prefix that we added in iframe-wrapper.ts
```
Maybe this will be more readable?

Reply:
Yeah
Good idea 💯